### PR TITLE
Training_script_refactor

### PR DESCRIPTION
**Pull request by jaidhyani**
This is a WIP refactor of the training script. It's not in a good state and there's a high probability I broke something while refactoring, but it has one redeeming quality: it runs. Does it run correctly and produce usable models? Excellent question, I have no idea yet.

From `src/`: ```python delphi/train/training.py```

### CODE REVIEW

**Comment by jaidhyani** [comment]
I don't think this is totally ready yet, but at this point I think it's probably worth it to (1) merge this into the `training-script` branch and (2) then merge the `training_script` branch into `main`.

File: src/delphi/train/architectures.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    arch = model_args["architecture"]
+    checkpoint_model_args = checkpoint["model_args"]
+    for k in args_to_load_from_checkpoint[arch]:
+        model_args[k] = checkpoint_model_args[k]
+    if arch == ModelTypes.LLAMA2C:
+        # create the model
+        gptconf = Llama2ModelArgs(**model_args)
+        model = Llama2Model(gptconf)
+        state_dict = checkpoint["model"]
+        # fix the keys of the state dictionary :(
```
Comment (SrGonao) [inline src/delphi/train/architectures.py:54]: Are we sure this still happens?

File: src/delphi/train/architectures.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        "dim",
+        "n_layers",
+        "n_heads",
+        "n_kv_heads",
+        "vocab_size",
+        "multiple_of",
+        "max_seq_len",
+    ],
+    ModelTypes.MAMBA: [
+        "n_layers",
```
Comment (SrGonao) [inline src/delphi/train/architectures.py:25]: Here also vocab_size should be set

File: src/delphi/train/gigaconfig.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    max_epochs=2,
+    eval_interval=500,
+    eval_iters=10,
+)
+
+
+def assert_config_sanity(config: GigaConfig):
+    assert config.vocab_source in ["llama2", "custom"]
+    assert (
+        config.vocab_source == "custom" or config.vocab_size == 32000
```
Comment (SrGonao) [inline src/delphi/train/gigaconfig.py]: do we want to keep 32K option or just the custom? I think just custom

File: src/delphi/train/iteration_params.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    num_steps: int
+    eval_iters: int
+    lr_decay_iters: int
+    tokens_per_iter: int
+
+
+def set_iteration_params(config, train_ds, validation_ds) -> IterationParams:
+    num_batches = len(train_ds) // config.batch_size
+    num_steps = num_batches // config.gradient_accumulation_steps
+    eval_iters = min(12, len(validation_ds) // config.batch_size)
```
Comment (SrGonao) [inline src/delphi/train/iteration_params.py:16]: what is the 12

Comment (jaidhyani) [reply src/delphi/train/iteration_params.py:16]: Something that was already there that I didn't want to change without understanding it first: https://github.com/delphi-suite/delphi/pull/31/files#diff-c113425bb7a4b6c38858b09ca918bc89bd243d139cdca30ab5dd386d9690935bR94

File: src/delphi/train/tokenized_chunks_dataset.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        ).to(self.device)
+        self.batched_tokens = tensor_tokens.flatten().unfold(
+            0, self.max_len + 1, self.max_len
+        )
+        self.indices = self._default_indices()
+
+    def _default_indices(self):
+        return list(range(len(self.batched_tokens)))
+
+    def shuffle(self, epoch: int):
```
Comment (SrGonao) [inline src/delphi/train/tokenized_chunks_dataset.py:35]: I don't know if this is correct. Shuffle alters the list in place, why are we changing the indices instead of the list?

File: src/delphi/train/architectures.py (hunk 1 above)
Comment (jaidhyani) [reply src/delphi/train/architectures.py:54]: No idea tbqh, that's just something I copied and never checked.

**Review by SrGonao — COMMENTED** [review]
I left a couple comments, only noticeable thing is on the shuffle, the rest is minor points and then mamba stuff that I should do

File: src/delphi/train/tokenized_chunks_dataset.py (hunk 5 above)
Comment (jaidhyani) [reply src/delphi/train/tokenized_chunks_dataset.py:35]: Idempotency. I want `shuffle(x)` to result in the same state regardless of when it's called - so not dependent on the state of the shuffling before it's called. By making it a deterministic shuffle of the default indices we can guarantee that it doesn't matter when you call `shuffle(x)`, you'll still get the same result.

Technically we don't need this for reproducibility if we commit to always calling shuffle with the same arguments in the same sequence, but I prefer to have it work this way to minimize the amount of state that needs to be kept track of when debugging - I don't want to have to repeat every shuffle in a given sequence to reproduce a problem.

File: src/delphi/constants.py
Comment (jannik-brinkmann) [inline src/delphi/constants.py:1]: seems good for now, but in the future we should consider adding the dataset as a parameter to the config

Comment (jaidhyani) [reply src/delphi/constants.py:1]: Could be both - a config param that defaults to a pre-defined constant.

File: src/delphi/train/iteration_params.py (hunk 4 above)
Comment (jannik-brinkmann) [reply src/delphi/train/iteration_params.py:16]: as far as I can tell this is a constant set by Karpathy to log at least every 12 batches.

File: src/delphi/train/tokenized_chunks_dataset.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+from torch.utils.data import Dataset
+
+from delphi.train.shuffle import shuffle_list
+
+
+class TokenizedChunksDataset(Dataset):
+    def __init__(self, tokenized_docs, max_seq_len, device):
+        self.device = device
+        self.tokenized_docs = tokenized_docs
+        self.doc_len = len(tokenized_docs[0]["tokens"])
```
Comment (jannik-brinkmann) [inline src/delphi/train/tokenized_chunks_dataset.py:11]: "doc_len" = "document_length" = context length?

File: src/delphi/train/tokenized_chunks_dataset.py (hunk 5 above)
Comment (jannik-brinkmann) [reply src/delphi/train/tokenized_chunks_dataset.py:35]: I like it that way

File: src/delphi/train/training.py [hunk 7]
```diff
@@ -1,331 +1,62 @@
...
-        if k.startswith(unwanted_prefix):
-            state_dict[k[len(unwanted_prefix) :]] = state_dict.pop(k)
-    model.load_state_dict(state_dict)
-    iter_num = checkpoint["iter_num"]
-    best_val_loss = checkpoint["best_val_loss"]
-model.to(device)
+model_training_state = load_model_training_state(config, device)
 
+# setup eval callbacks
+eval_callbacks = [save_checkpoint_if_needed]
```
Comment (jannik-brinkmann) [inline src/delphi/train/training.py:41]: maybe I am missing something, but if eval_callbacks only includes saving the checkpoint, where does it log to wandb?

Comment (jannik-brinkmann) [reply src/delphi/train/training.py:41]: nvm, found it in line 44 lol

**Review by jannik-brinkmann — APPROVED** [review]
looks good to me. as discussed in Discord, I will start a training run tomorrow after the PR is merged to see whether the performance matches our current models

File: src/delphi/train/tokenized_chunks_dataset.py (hunk 6 above)
Comment (jaidhyani) [reply src/delphi/train/tokenized_chunks_dataset.py:11]: Document length, but I don't think that's context length. `max_seq_len` is model context length - or more accurately, how long training samples are.

File: src/delphi/train/tokenized_chunks_dataset.py (hunk 5 above)
Comment (SrGonao) [reply src/delphi/train/tokenized_chunks_dataset.py:35]: I just didnt see where you then apply the indices to the list, but if you do it somewhere I trust you
